compilers/concrete-compiler/compiler/lib/Runtime/simulation.cpp

```diff
@@ -183,16 +205,147 @@ void sim_wop_pbs_crt(
 
 uint64_t sim_neg_lwe_u64(uint64_t plaintext) { return ~plaintext + 1; }
 
+uint64_t sim_add_lwe_u64(uint64_t lhs, uint64_t rhs, char *loc,
+                         bool is_signed) {
+  const char msg_f[] =
+      "WARNING at %s: overflow happened during addition in simulation\n";
+
+  uint64_t result = lhs + rhs;
+
+  if (is_signed) {
+    // We shift left to discard the padding bit and only consider the message
+    // for easier overflow checking
+    int64_t lhs_signed = (int64_t)lhs << 1;
+    int64_t rhs_signed = (int64_t)rhs << 1;
+    if (lhs_signed > 0 && rhs_signed > INT64_MAX - lhs_signed)
+      printf(msg_f, loc);
+    else if (lhs_signed < 0 && rhs_signed < INT64_MIN - lhs_signed)
+      printf(msg_f, loc);
+  } else if (lhs > UINT63_MAX - rhs || result > UINT63_MAX) {
+    printf(msg_f, loc);
+  }
+  return result;
+}
+
+uint64_t sim_mul_lwe_u64(uint64_t lhs, uint64_t rhs, char *loc,
+                         bool is_signed) {
+  const char msg_f[] =
+      "WARNING at %s: overflow happened during multiplication in simulation\n";
+
+  uint64_t result = lhs * rhs;
+
+  if (is_signed) {
+    // We shift left to discard the padding bit and only consider the message
+    // for easier overflow checking
+    int64_t lhs_signed = (int64_t)lhs << 1;
+    int64_t rhs_signed = (int64_t)rhs << 1;
+    if (lhs_signed != 0 && rhs_signed > INT64_MAX / lhs_signed)
+      printf(msg_f, loc);
+    else if (lhs_signed != 0 && rhs_signed < INT64_MIN / lhs_signed)
+      printf(msg_f, loc);
+  } else if (rhs != 0 && (lhs > UINT63_MAX / rhs || result > UINT63_MAX)) {
+    printf(msg_f, loc);
+  }
+  return result;
+}
+
+// a copy of memref_encode_expand_lut_for_bootstrap but which encodes overflow
+// and sign info into the LUT. Those information should later be discarder by
+// the LUT function
 void sim_encode_expand_lut_for_boostrap(
-    uint64_t *out_allocated, uint64_t *out_aligned, uint64_t out_offset,
-    uint64_t out_size, uint64_t out_stride, uint64_t *in_allocated,
-    uint64_t *in_aligned, uint64_t in_offset, uint64_t in_size,
-    uint64_t in_stride, uint32_t poly_size, uint32_t output_bits,
-    bool is_signed) {
-  return memref_encode_expand_lut_for_bootstrap(
-      out_allocated, out_aligned, out_offset, out_size, out_stride,
-      in_allocated, in_aligned, in_offset, in_size, in_stride, poly_size,
-      output_bits, is_signed);
+    uint64_t *output_lut_allocated, uint64_t *output_lut_aligned,
```

Comment:
It could be good to not duplicate the code of `memref_encode_expand_lut_for_bootstrap`, you can have a closure or something like that, that patch the lut encoding.
For the signed information I guess it not be mandatory to be encoded in the lut but could be passed as an argument in `sim_bootstrap_lwe_u64` (but that is equivalent so as you want).

Reply:
I didn't intend to copy it, but as you can see, encoding info into the LUT isn't as straightforward as iterating through a vector, you have duplicates of the same value centered over an index, and while trying to implement the encoding before making the call to the memref_* function, I realized that I added most of it, so it's simpler IMO to duplicate the function.